lib/tree.js

```diff
@@ -12,47 +12,132 @@ class Tree {
         this.root = null;
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+    add_helper(current, new_node) {
+        if (current == null) { return new_node }
+
+        if (new_node.key <= current.key) { //go left
+            current.left = this.add_helper(current.left, new_node)
+        } else {
+            current.right = this.add_helper(current.right, new_node)
+        }
+        return current
+    }
+
+    // Time Complexity: O(n)
+    // Space Complexity: O(1)
     add(key, value) {
```

Comment:
👍 , but time & space are O(log n) if the tree is balanced and O(n) if the tree is not balanced.  

You have space complexity because the helper method are recursive.